### PR TITLE
Create FrameIterator type

### PR DESCRIPTION
**Pull request by Headline22**
I decided to redact these changes from the LogStackTrace pull request since these are unrelated.

If we decide to add a stack frame iterator for sourcemod plugins, this is what that implementation would look like. Comments & criticisms are welcome.

### CODE REVIEW

File: plugins/include/sourcemod.inc [hunk 1]
```diff
@@ -645,6 +645,40 @@ native int LoadFromAddress(Address addr, NumberType size);
...
 native void StoreToAddress(Address addr, int data, NumberType size); 
 
+methodmap FrameIterator < Handle {
+	// Creates a stack frame iterator to build your own stack traces.
+	// @return New handle to a FrameIterator.
+	public native FrameIterator();
+	
+	// Determines if there are any more stack frames to iterate.
+	// @return True if there are no more stack frames left.
+	public native bool Done();
```
Comment (KyleSanderson) [inline plugins/include/sourcemod.inc]: I don't think this is natural, Next should return true or false. Done can be done away with completely (I think), or you can keep it to match SQL_MoreRows.

https://github.com/alliedmodders/sourcemod/blob/750767289558236da6c3554c734745aef35f58dd/core/logic/smn_database.cpp#L1062

File: core/logic/smn_core.cpp [hunk 2]
```diff
@@ -780,6 +792,163 @@ static cell_t IsNullString(IPluginContext *pContext, const cell_t *params)
...
+	if (!filePath)
+	{
+		return pContext->ThrowNativeError("Iterator out of bounds!");
+	}
+
+	char* buffer;
+	pContext->LocalToString(params[2], &buffer);
+	size_t size = static_cast<size_t>(params[3]);
+	
+	snprintf(buffer, size, "%s", filePath);
```
Comment (KyleSanderson) [inline core/logic/smn_core.cpp]: ke::SafeStrcpy

File: core/logic/FrameIterator.h [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+class SafeFrameIterator
+{
+public:
+
+	/**
+	 * Struct which holds all of the cached values for each individual frame.
+	 */
+	struct FrameInfo
+	{
+		unsigned LineNumber;
```
Comment (KyleSanderson) [inline core/logic/FrameIterator.h]: Typically you want larger constructs to be at the top for packing. I'm not sure if these are going to be promoted or not but I would half expect them to be.

File: core/logic/smn_core.cpp [hunk 4]
```diff
@@ -780,6 +792,163 @@ static cell_t IsNullString(IPluginContext *pContext, const cell_t *params)
...
+
+	if ((err=handlesys->ReadHandle(hndl, g_FrameIter, &sec, (void **)&iterator)) != HandleError_None)
+	{
+		return pContext->ThrowNativeError("Could not read Handle %x (error %d)", hndl, err);
+	}
+	
+	const char* filePath = iterator->FilePath();
+	if (!filePath)
+	{
+		return pContext->ThrowNativeError("Iterator out of bounds!");
```
Comment (KyleSanderson) [inline core/logic/smn_core.cpp]: It's good you're checking these, but I would try and give a little more information if possible. I'm not sure the '!' is present anywhere else in SM, so matching style would be an improvement.

File: core/logic/FrameIterator.cpp [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+		it->Next(); 
+	}
+	
+	it->Reset();
+	current = 0;
+}
+
+bool SafeFrameIterator::Done() const
+{
+	return current >= (frames.length() - 1);
```
Comment (KyleSanderson) [inline core/logic/FrameIterator.cpp]: current < frames.length()

File: core/logic/smn_core.cpp [hunk 6]
```diff
@@ -91,13 +95,21 @@ class CoreNativeHelpers :
 	}
 	void OnHandleDestroy(HandleType_t type, void *object)
 	{
-		IPluginIterator *iter = (IPluginIterator *)object;
-		iter->Release();
+		if (type == g_FrameIter)
```
Comment (KyleSanderson) [inline core/logic/smn_core.cpp]: maybe use switch here, there's a mixed style throughout SM.

File: core/logic/smn_core.cpp [hunk 7]
```diff
@@ -780,6 +792,163 @@ static cell_t IsNullString(IPluginContext *pContext, const cell_t *params)
...
+	sec.pIdentity = g_pCoreIdent;
+	sec.pOwner = pContext->GetIdentity();
+
+	if ((err=handlesys->ReadHandle(hndl, g_FrameIter, &sec, (void **)&iterator)) != HandleError_None)
+	{
+		return pContext->ThrowNativeError("Could not read Handle %x (error %d)", hndl, err);
+	}
+	
+	const char* functionName = iterator->FunctionName();
+
```
Comment (KyleSanderson) [inline core/logic/smn_core.cpp]: crunch here, you're doing it below.

File: core/logic/FrameIterator.cpp [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+void SafeFrameIterator::Next()
+{
+	current++;
```
Comment (KyleSanderson) [inline core/logic/FrameIterator.cpp]: This should not increment endlessly, after you hit Done this should stop.

File: core/logic/smn_core.cpp (hunk 6 above)
Comment (Headline22) [reply core/logic/smn_core.cpp]: Compiler demands that g_FrameIter and the likes must be constant variables. Errors specifically are
```
core\logic\smn_core.cpp(100): error C2131: expression did not evaluate to a constant
core\logic\smn_core.cpp(100): note: failure was caused by non-constant arguments or reference to a non-constant symbol
core\logic\smn_core.cpp(100): note: see usage of 'g_FrameIter'
core\logic\smn_core.cpp(105): error C2131: expression did not evaluate to a constant
core\logic\smn_core.cpp(105): note: failure was caused by non-constant arguments or reference to a non-constant symbol
core\logic\smn_core.cpp(105): note: see usage of 'g_PlIter'
core\logic\smn_core.cpp(100): error C2051: case expression not constant
core\logic\smn_core.cpp(105): error C2051: case expression not constant
```

Comment (KyleSanderson) [reply core/logic/smn_core.cpp]: Ah, yes you're (well, msvc is) absolutely right.

File: core/logic/FrameIterator.cpp [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+bool SafeFrameIterator::Done() const
+{
+	return current == frames.length();
+}
+
+bool SafeFrameIterator::Next()
+{
+	if(!this->Done())
```
Comment (KyleSanderson) [inline core/logic/FrameIterator.cpp]: style

File: core/logic/FrameIterator.cpp [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+ * or <http://www.sourcemod.net/license.php>.
+ *
+ * Version: $Id$
+ */
+ 
+#include "FrameIterator.h"
+
+SafeFrameIterator::SafeFrameIterator(IFrameIterator *it)
+{	
+	while (!it->Done())
```
Comment (KyleSanderson) [inline core/logic/FrameIterator.cpp]: style, but do {} while(it->Next());

Comment (KyleSanderson) [reply core/logic/FrameIterator.cpp]: actually if you do this you'd have to check it->Done() first; so it's all good.

Comment (asherkin) [reply core/logic/FrameIterator.cpp]: There could be 0 frames to iterate.

Comment (KyleSanderson) [reply core/logic/FrameIterator.cpp]: That's not possible for the Pawn API.

**Pull request merged**